### PR TITLE
Support suda to persist undo (#25)

### PR DESCRIPTION
**Pull request by kevinhwang91**
Write the undo file manually.
Only test under *unix.

### CODE REVIEW

**Comment by lambdalisue** [comment]
I need to investigate it carefully so please be patient 👍

File: autoload/suda.vim [hunk 1]
```diff
@@ -266,6 +270,27 @@ function! s:prefix_searchpattern() abort
...
+    " TODO
+    " I don't have Windows environment. But I know that path separators in
+    " Windows are both ':' and '\' for undo file.
+  else
+    let p = resolve(a:path)
+    if &undodir == '.'
+      let undo_path = printf('%s/.%s.un~', fnamemodify(p, ':p:h'),
+                  \ fnamemodify(p, ':t'))
+    else
+      let undo_path = printf('%s/%s', &undodir, fnamemodify(p, ':p:gs?/?\\%?'))
```
Comment (kevinhwang91) [inline autoload/suda.vim]: Maybe I should use `fnameescape()`  to escape `%`

**Comment by lambdalisue** [comment]
I'm sorry for the late this. Well, what is the ownership and permission of the undofile? I think it must be exactly equal to the file which suda read otherwise others may able to read without permission right?

**Comment by kevinhwang91** [comment]
`:h undodir`

> *'undodir'* *'udir'* *E5003*
'undodir' 'udir'	string	(default "$XDG_DATA_HOME/nvim/undo")
			global
	List of directory names for undo files, separated with commas.
	See |'backupdir'| for details of the format.
	"." means using the directory of the file.  The undo file name for
	"file.txt" is ".file.txt.un~".
	For other directories the file name is the full path of the edited
	file, with path separators replaced with "%".
	When writing: The first directory that exists is used.  "." always
	works, no directories after "." will be used for writing.  If none of 
	the directories exist Neovim will attempt to create last directory in 
	the list.
	When reading all entries are tried to find an undo file.  The first
	undo file that exists is used.  When it cannot be read an error is
	given, no further entry is used.
	See |undo-persistence|.
	This option cannot be set from a |modeline| or in the |sandbox|, for
	security reasons.


if user set `undodir` to `path`, running suda without permission for undofile, neovim will complain 
```
Error detected while processing function suda#BufWriteCmd[3]..suda#write[46]..<SNR>59_write_undo:
line   18:
E828: Cannot open undo file for writing: xxxx.un~
```
otherwise, the undofile are always stored in `$XDG_DATA_HOME/nvim/undo` directory, which is the user home subpath.

No matter what the `undodir` value is,  the permission of undofile is the same as the parent process of neovim.

We can simply try catch the error or ignore the complaint if we want.

**Comment by lambdalisue** [comment]
I'm not professional on this site thus this is a question but what happens if

- Alan is sudoers thus he can read `/secret`
- Bob is NOT sudoers but he belongs `alan` group
- Alan read `/secret` by suda and `/home/.config/nvim/undo/secret` has written

In above case, can Bob read `/home/.config/nvim/undo/secret`? Or is my presupportion wrong?

**Comment by kevinhwang91** [comment]
I am not sure what you care about.
`$XDG_DATA_HOME/nvim/undo` is extended to `/home/kevin/.local/share/nvim/undo` in my system, `kevin` is my user name.
This path is absolutely owned by the user.

If the user sets the `undodir` to the path which without permission, not only we can't manually write undofile, but also neovim itself will not be able to write the undofile.

**Comment by lambdalisue** [comment]
> $XDG_DATA_HOME/nvim/undo is extended to /home/kevin/.local/share/nvim/undo in my system, kevin is my user name.
This path is absolutely owned by the user.

Bob belongs to the `alan` group implied that Bob can read `alan`'s files in my explanation. Thus the file existing in user's home directory is not enough for safe reason in my point of view.

> If the user sets the undodir to the path which without permission, not only we can't manually write undofile, but also neovim itself will not be able to write the undofile.

Opposite. I worry about if `undodir` is permission-less like 777. In that case, I think everyone can read files written by suda.vim's persistent undo feature which, for example, undofile of `/etc/shadow`.

Alacritty 2020-08-06 02-25-52

For example, the above image is a result of `ls -al` in my `undodir` on macOS. It said that permissions of undo-files are 644. It means even non-staff users can read those files if they can reach the `undodir`, right? 
This is why I worry about this PR. So please explain why this is safe.

**Comment by kevinhwang91** [comment]
If you concern the undofile leak the password, we can change the mode of undofile written by suda to `600`, group and others can't read the file, but the user can go on using this undofile in neovim.

**Comment by lambdalisue** [comment]
I think about this and I believe that this PR can be a security hole.

Undo files produced by suda **MUST** have exactly same owner, group, and permissions as **the original file**. Otherwise, cracker can read secret contents through undo files produced by suda **WITHOUT** being super user

**Comment by lambdalisue** [comment]
> If you concern the undofile leak the password, we can change the mode of undofile written by suda to 600, group and others can't read the file, but the user can go on using this undofile in neovim.

That's not enough. I'm talking about cases even the user cannot read the content without being superuser. Thus the permission level must be exactly same with the original file (and I don't think it's handy enough)

**Comment by kevinhwang91** [comment]
The same permissions will inevitably cause undo to fail to be read from neovim normally.

With using undofile, sensitive information must be left on undo.

Maybe we should leave the choice to the user, give an option disabled by default, and point out the risk of using that option.

**Comment by kevinhwang91** [comment]
The same permissions will inevitably cause failure about reading undofile from neovim normally.

With using undofile, sensitive information must be left on undo.

Maybe we should leave the choice to the user, give an option disabled by default, and point out the risk of using that option.

**Comment by lambdalisue** [comment]
I'm really sorry but it's not acceptable even the feature has disabled by default. There is too sensitive data in the world and I don't want my plugin to be a security hole. 

Thanks for your patient contributions and I'm sorry again for the rejection.
Feel free to keep using your fork by your own risk.

**Comment by sorz** [comment]
I think it's depend on the specific system configuration. It's prefect to have 600 on user's home directory so that only the user (who is sudoer) can access the undo files, whatever the permission on these undo files themself. There may be a concern that the user now no longer need the sudo password to read the undo file. But again, it depend on the configuration, sudoer may not need password to get root permission from the beginning. Moreover, nowadays many computers are single-user use (like my laptop).

I'm appreciate to see the author is so careful about security. But I also think this feature is valid and not-so-danger on some cases.

**Comment by lambdalisue** [comment]
I see... Maybe I am overthinking things, but I am not sure, so I am forced to be cautious. I'm starting to think that we could offer the functionality on an opt-in basis as suggested and write a warning in the help, but I'm not sure and it's a bit difficult to write a warning.

**Pull request closed without merging**